node/src/components/storage/disjoint_sequences.rs

```diff
@@ -0,0 +1,317 @@
+pub(super) struct Sequence {
+    /// The upper bound (inclusive) of the sequence.
+    high: u64,
+    /// The lower bound (inclusive) of the sequence.
+    low: u64,
+}
+
+impl Sequence {
+    /// Constructs a new sequence containing only `value`.
+    fn new(value: u64) -> Self {
+        Sequence {
+            high: value,
+            low: value,
+        }
+    }
+
+    #[cfg(test)]
+    pub(crate) fn new_with_bounds(low: u64, high: u64) -> Self {
+        Sequence { high, low }
+    }
+
+    /// Tries to insert `value` into the sequence.
+    ///
+    /// Returns an outcome which indicates where the value was inserted if at all.
+    fn try_insert(&mut self, value: u64) -> InsertOutcome {
+        if value == self.high + 1 {
+            self.high = value;
+            InsertOutcome::ExtendedHigh
+        } else if value >= self.low && value <= self.high {
+            InsertOutcome::AlreadyInSequence
+        } else if value + 1 == self.low {
+            self.low = value;
+            InsertOutcome::ExtendedLow
+        } else if value > self.high {
+            InsertOutcome::TooHigh
+        } else {
+            InsertOutcome::TooLow
+        }
+    }
+}
+
+/// Represents a collection of disjoint sequences of `u64`s.
+///
+/// The collection is kept ordered from high to low, and each entry represents a discrete portion of
+/// the space from [0, u64::MAX] with a gap of at least 1 between each.
+///
+/// The collection is ordered this way to optimize insertion for the normal use case: adding
+/// monotonically increasing values representing the latest block height.
+///
+/// As values are inserted, if two separate sequences become contiguous, they are merged into a
+/// single sequence.
+///
+/// For example, if `sequences` contains `[9,9], [7,3]` and `8` is inserted, then `sequences` will
+/// be reduced to `[9,3]`.
+#[derive(Default, Debug, DataSize)]
+pub(super) struct DisjointSequences {
+    sequences: Vec<Sequence>,
+}
+
+impl DisjointSequences {
+    /// Inserts `value` into the appropriate sequence and merges sequences if required.
+    ///
+    /// Note, this method is efficient where `value` is one greater than the current highest value.
+    /// However, it's not advisable to use this method in a loop to rebuild a `DisjointSequences`
+    /// from a large collection of randomly-ordered values.  In that case, it is very much more
+    /// efficient to use `DisjointSequences::from(mut input: Vec<u64>)`.
+    pub(super) fn insert(&mut self, value: u64) {
+        let mut iter_mut = self.sequences.iter_mut().enumerate().peekable();
+
+        // The index at which to add a new `Sequence` containing only `value`.
+        let mut maybe_insertion_index = Some(0);
+        // The index of a `Sequence` to be removed due to the insertion of `value` causing two
+        // consecutive sequences to become contiguous.
+        let mut maybe_removal_index = None;
+        while let Some((index, sequence)) = iter_mut.next() {
```

Comment:
Since the vector of sequences is ordered, this could be done with a binary search, couldn't it?

Have a function on `Sequence`:
```rust
fn in_sequence(&self, value: u64) -> Ordering {
    if value >= self.low && value <= self.high {
        Ordering::Equal
    } else if value < self.low {
        Ordering::Less
    } else {
        Ordering::Greater
    }
}
```

Then, call `self.sequences.binary_search_by(|seq| seq.in_sequence(value))` - if that returns `Ok(index)`, value is already in a sequence. If it returns `Err(index)`, then we have to try adding it to sequences at `index` and, if that fails, `index+1`. If one of them gets extended, good, if not, we need to create a new one at `index`.

This should improve performance significantly for very fragmented chains (which might not be a common case, but it could be still worth it?).

Reply:
I think this might be a premature optimisation.  I agree for the fragmented case, but I think our use case will bring in a lot of entries which will fall into the first sequence in the disjoint collection.  So binary searching might actually be a pessimisation?

Either way, I think I'd prefer to leave this as is for now and we can look to benchmark and optimise later if there's a need.